editor/inserter/menu.js

```diff
@@ -58,37 +57,69 @@ class InserterMenu extends Component {
 		} );
 	}
 
-	selectBlock( name ) {
+	selectBlock( blockType ) {
 		return () => {
-			this.props.onSelect( name );
+			this.props.onSelect( blockType.name );
 			this.setState( {
 				filterValue: '',
 				currentFocus: null,
 			} );
 		};
 	}
 
-	getVisibleBlocks( blockTypes ) {
-		return filter( blockTypes, this.isShownBlock );
+	applySearchFilter( blockTypes ) {
+		const matchesSearch = ( block ) => block.title.toLowerCase().indexOf( this.state.filterValue.toLowerCase() ) !== -1;
+		return filter( blockTypes, matchesSearch );
+	}
+
+	getBlocksForCurrentTab() {
+		// if we're searching, use everything, otherwise just get the blocks visible in this tab
+		if ( this.state.filterValue ) {
+			return getBlockTypes();
+		}
+		switch ( this.state.tab ) {
+			case 'recent':
+				return this.props.recentlyUsedBlocks;
+			case 'blocks':
+				return filter( getBlockTypes(), ( block ) => block.category !== 'embed' );
+			case 'embeds':
+				return filter( getBlockTypes(), ( block ) => block.category === 'embed' );
+		}
 	}
 
-	sortBlocksByCategory( blockTypes ) {
+	sortBlocks( blockTypes ) {
+		if ( 'recent' === this.state.tab && ! this.state.filterValue ) {
+			return blockTypes;
+		}
+
 		const getCategoryIndex = ( item ) => {
 			return findIndex( getCategories(), ( category ) => category.slug === item.category );
 		};
 
 		return sortBy( blockTypes, getCategoryIndex );
 	}
 
+	addRecentBlocks( blocksByCategory ) {
```

Comment:
Just to clarify why this needs a separate step, 'recent' isn't a category because we don't have the ability for a block to be in two categories, so the method that sorts blocks by category has no knowledge of 'recent', and this method is adding the recent category to the object that holds the blocks grouped by category.